pkg/ctl/cmdutils/nodegroup_filter.go

```diff
@@ -0,0 +1,134 @@
+	return &NodeGroupFilter{
+		IgnoreAllExisting: true,
+
+		existing: sets.NewString(),
+	}
+}
+
+// ApplyOnlyFilter parses given globs for exclusive filtering
+func (f *NodeGroupFilter) ApplyOnlyFilter(globExprs []string, cfg *api.ClusterConfig) error {
+	for _, expr := range globExprs {
+		compiledExpr, err := glob.Compile(expr)
+		if err != nil {
+			return errors.Wrapf(err, "parsing glob filter %q", expr)
+		}
+		f.only = append(f.only, compiledExpr)
+	}
+	f.onlySpec = strings.Join(globExprs, ",")
+	return f.onlyFilterMatchesAnything(cfg)
+}
+
+func (f *NodeGroupFilter) onlyFilterMatchesAnything(cfg *api.ClusterConfig) error {
+	if len(f.only) == 0 {
+		return nil
+	}
+	for i := range cfg.NodeGroups {
+		for _, g := range f.only {
+			if g.Match(cfg.NodeGroups[i].Name) {
+				return nil
+			}
+		}
+	}
+	return fmt.Errorf("no nodegroups match filter specification: %q", f.onlySpec)
+}
+
+// ApplyExistingFilter uses stackManager to list existing nodegroup stacks and configures
+// the filter accordingly
+func (f *NodeGroupFilter) ApplyExistingFilter(stackManager *manager.StackCollection) error {
+	if !f.IgnoreAllExisting {
+		return nil
+	}
+
+	existing, err := stackManager.ListNodeGroupStacks()
+	if err != nil {
+		return err
+	}
+
+	f.existing.Insert(existing...)
+
+	return nil
+}
+
+// Match checks given nodegroup against the filter
+func (f *NodeGroupFilter) Match(ng *api.NodeGroup) bool {
+	if f.IgnoreAllExisting && f.existing.Has(ng.Name) {
+		return false
+	}
+
+	for _, g := range f.only {
+		if g.Match(ng.Name) {
+			return true // return first match
+		}
+	}
+
+	// if no globs were given, match everything,
+	// if false - we haven't matched anything so far
+	return len(f.only) == 0
+}
+
+// MatchAll checks all nodegroups against the filter and returns all of
+// matching names as set
+func (f *NodeGroupFilter) MatchAll(cfg *api.ClusterConfig) sets.String {
+	names := sets.NewString()
+	for _, ng := range cfg.NodeGroups {
+		if f.Match(ng) {
+			names.Insert(ng.Name)
+		}
+	}
+	return names
+}
+
+// LogInfo prints out a user-friendly message about how filter was applied
+func (f *NodeGroupFilter) LogInfo(cfg *api.ClusterConfig) {
+	count := f.MatchAll(cfg).Len()
+	filteredOutCount := len(cfg.NodeGroups) - count
+	if filteredOutCount > 0 {
+		reasons := []string{}
+		if f.onlySpec != "" {
+			reasons = append(reasons, fmt.Sprintf("--only=%q was given", f.onlySpec))
+		}
+		if existingCount := f.existing.Len(); existingCount > 0 {
+			reasons = append(reasons, fmt.Sprintf("%d nodegroup(s) (%s) already exist", existingCount, strings.Join(f.existing.List(), ", ")))
+		}
+		logger.Info("%d nodegroup(s) were filtered out: %s", filteredOutCount, strings.Join(reasons, ", "))
+	}
+}
+
+// CheckEachNodeGroup iterates over each nodegroup and calls check function if it matches the filter
+func (f *NodeGroupFilter) CheckEachNodeGroup(nodeGroups []*api.NodeGroup, check func(i int, ng *api.NodeGroup) error) error {
```

Comment:
@martina-if I though it's a bit better to make it a method on `NodeGroupFilter` for now, so we don't have to pass `ngFilter` explicitly each time...